nx/lib/nx.ex

```diff
@@ -6658,6 +6658,251 @@ defmodule Nx do
+      f64
+      5.0
+    >
+
+    iex> Nx.norm(Nx.tensor([[3, 0], [0, -4]]))
+    #Nx.Tensor<
+      f64
+      5.0
+    >
+
+    iex> Nx.norm(Nx.tensor([[3, 4], [0, -4]]), axes: [1], keep_axes: true)
+    #Nx.Tensor<
+      f64[2][1]
+      [
+        [5.0],
+        [4.0]
+      ]
+    >
+
+    iex> Nx.norm(Nx.tensor([0, 1]), my_opt: 1)
+    ** (RuntimeError) unknown key :my_opt in [my_opt: 1], expected one of [:ord, :axes, :keep_axes]
+
+  ### Caveats
+
+  For big values of p, f64 rounding errors come into play
+  """
+
+  def norm(t_raw, input_opts \\ []) when is_list(input_opts) do
+    t = tensor!(t_raw)
+    assert_keys!(input_opts, [:ord, :axes, :keep_axes])
+
+    default_axes_opts = [axes: nil, keep_axes: false, ord: nil]
+    opts = Keyword.merge(default_axes_opts, input_opts)
+
+    case shape(t) do
+      {_, _} ->
+        :ok
+
+      {_} ->
+        :ok
+
+      s ->
+        raise "expected 1-D or 2-D tensor. Received a tensor with shape #{inspect(s)} instead."
+    end
+
+    rank = rank(t)
+    ord = get_norm_ord!(opts, rank)
+
+    axes_opts = Keyword.take(opts, [:axes, :keep_axes])
+
+    do_p_norm(t, ord, rank, axes_opts)
+  end
+
+  defp get_norm_ord!(opts, dims) do
+    ord = Keyword.get(opts, :ord)
+
+    case {ord, dims} do
+      {nil, _} ->
+        2
+
+      {:frobenius, 2} ->
+        2
+
+      {:frobenius, _} ->
+        raise "expected a 2-D tensor. Got a #{dims}-D tensor."
+
+      {:nuclear, _} ->
+        raise "nuclear norm not implemented yet."
+
+      {ord, 2} when ord in [:inf, :neg_inf] ->
+        ord
+
+      {ord, 1} when ord in [:inf, :neg_inf] ->
+        ord
+
+      {0, 2} ->
+        raise "ord #{inspect(ord)} not implemented for 2-D tensor."
+
+      {0, 1} ->
+        0
+
+      {ord, _} ->
+        ord
+    end
+  end
+
+  defp do_p_norm(t, 0, 1, opts) do
+    t
+    |> equal(0)
+    |> logical_not()
+    |> sum(opts)
+  end
+
+  defp do_p_norm(_t, 0, 2, _opts) do
+    raise "expected 1-D tensor for ord: 0. Got: 2-D tensor."
+  end
+
+  defp do_p_norm(t, ord, 2, axes_opts) when ord in [:inf, :neg_inf] do
+    opts = Keyword.merge(axes_opts, axes: [1])
+
+    function = if ord == :inf, do: &reduce_max/1, else: &reduce_min/1
+
+    t
+    |> Nx.abs()
+    |> sum(opts)
+    |> function.()
+  end
+
+  defp do_p_norm(t, ord, 1, axes_opts) when ord in [:inf, :neg_inf] do
+    function = if ord == :inf, do: &reduce_max/2, else: &reduce_min/2
+
+    t
+    |> Nx.abs()
+    |> function.(axes_opts)
+  end
+
+  defp do_p_norm(t, ord, 2, opts) when ord in [1, -1] do
+    function =
+      if ord == 1 do
+        &reduce_max/2
+      else
+        &reduce_min/2
+      end
+
+    t
+    |> Nx.abs()
+    |> sum(axes: [0])
+    |> function.(Keyword.take(opts, [:keep_axes]))
+  end
+
+  defp do_p_norm(_t, ord, 2, _opts) when ord not in -1..2 do
+    raise "invalid ord for 2-D tensor. Got: #{inspect(ord)}"
+  end
+
+  defp do_p_norm(%{type: type} = t, ord, _, opts) do
+    inv_ord = divide(1, ord)
```

Comment:
I would be careful using `divide` here because it might cause unnecessary type promotion later on. Instead I would do:

```elixir
inv_ord = tensor(1 / ord, type: type)
```

Also, is it possible `ord` is 0? Maybe it was caught earlier, but if not, be careful of division by 0 :)

Reply:
There are special function clauses for `ord: 0`. The numpy docs specified it in a similar manner to what I had in the original draft: returns the count of non-zero elements in the tensor

Reply:
I ran into some problems using `type: type` because `type` can represent an integer type. I've refactored this line to `tensor(1/p)` instead